src/SonarQube.Client/Services/SonarQubeService2.cs

```diff
@@ -0,0 +1,205 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using SonarQube.Client.Api;
+using SonarQube.Client.Api.Requests;
+using SonarQube.Client.Helpers;
+using SonarQube.Client.Messages;
+using SonarQube.Client.Models;
+
+namespace SonarQube.Client.Services
+{
+    public class SonarQubeService2 : ISonarQubeService
+    {
+        internal const int MaximumPageSize = 500;
+        internal static readonly Version OrganizationsFeatureMinimalVersion = new Version(6, 2);
+
+        private readonly HttpMessageHandler messageHandler;
+        private readonly RequestFactory requestFactory;
+
+        private HttpClient httpClient;
+
+        private Version sonarQubeVersion = null;
+
+        private bool connected;
+
+        public bool HasOrganizationsFeature
+        {
+            get
+            {
+                EnsureIsConnected();
+
+                return sonarQubeVersion >= OrganizationsFeatureMinimalVersion;
+            }
+        }
+
+        public bool IsConnected => connected;
+
+        public SonarQubeService2(HttpMessageHandler messageHandler, RequestFactory requestFactory)
+        {
+            this.messageHandler = messageHandler;
+            this.requestFactory = requestFactory;
+        }
+
+        private Task<TResponse> InvokeRequestAsync<TRequest, TResponse>(CancellationToken token)
+            where TRequest : IRequest<TResponse>
+        {
+            return InvokeRequestAsync<TRequest, TResponse>(request => { }, token);
+        }
+
+        private async Task<TResponse> InvokeRequestAsync<TRequest, TResponse>(Action<TRequest> configure,
+            CancellationToken token)
+            where TRequest : IRequest<TResponse>
+        {
+            EnsureIsConnected();
+
+            var request = CreateRequest<TRequest>();
+            configure(request);
+
+            var result = await request.InvokeAsync(httpClient, token);
+
+            return result;
+        }
+
+        private TRequest CreateRequest<TRequest>()
+            where TRequest : IRequest
+        {
+            return requestFactory.Create<TRequest>(sonarQubeVersion);
+        }
+
+        public async Task ConnectAsync(ConnectionInformation connection, CancellationToken token)
+        {
+            httpClient = new HttpClient(messageHandler)
+            {
+                BaseAddress = connection.ServerUri,
```

Comment:
Once a request was made, BaseAddress cannot be changed anymore, hence we create a new HttpClient for each connection